src/main/java/frc/robot/subsystems/DriveSubsystem.java

```diff
@@ -0,0 +1,124 @@
+
+        this.field2d = new Field2d();
+
+        differentialDriveOdometry = new DifferentialDriveOdometry(
+                new Rotation2d(getAngleDegrees()),
+                getLeftDistancePassedMeters(),
+                getRightDistancePassedMeters()
+        );
+        field2d.setRobotPose(differentialDriveOdometry.getPoseMeters().getX(), differentialDriveOdometry.getPoseMeters().getY(), differentialDriveOdometry.getPoseMeters().getRotation());
+
+        initialize();
+
+
+    }
+
+    public Field2d getField2d() {
+        return field2d;
+    }
+
+    public double getLeftDistancePassedMeters() {
+        return leftFrontMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public double getRightDistancePassedMeters() {
+        return rightBackMotor.getSelectedSensorPosition() / RobotMap.TALON_ENCODER_PPR * RobotMap.DRIVE_WHEEL_RADIUS * 2 * Math.PI;
+    }
+
+    public void initialize() {
+        pigeon2.reset();
+        leftFrontMotor.setSelectedSensorPosition(0);
+        rightBackMotor.setSelectedSensorPosition(0);//inits
+    }
+
+    public double getAngleDegrees() {
+        return (360 - pigeon2.getAngle()) % 360;// the value returned will be from 0 - 360 depending on its location
+        // 0  - degree after initializing, 359 - one degree to the right, 1 - one degree to the left// 90 will be 90 degrees to the left
+    }
+
+    public void powerLeftMotors(double powerL) {
+        leftFrontMotor.set(powerL);
+        leftBackMotor.set(powerL);
+    }
+
+    public void powerRightMotors(double powerR) {
+        rightFrontMotor.set(powerR);
+        rightBackMotor.set(powerR);
+    }
+
+    public void stop() {
+        rightBackMotor.stopMotor();
+        rightFrontMotor.stopMotor();
+        leftBackMotor.stopMotor();
+        leftFrontMotor.stopMotor();
+    }
+
+    //public void setOdometryPose2d (Pose2d pose2d){
+    //   differentialDriveOdometry.update(pose2d.getRotation(), getRightDistancePassedMeters(), getLeftDistancePassedMeters());
+    //}
+
+    private void updateOdometry() {
+        differentialDriveOdometry.update(
+                pigeon2.getRotation2d(),
+                getLeftDistancePassedMeters(),
+                getRightDistancePassedMeters()
+        );
+    }
+
+    public void periodic() {
+        differentialDriveOdometry.update(new Rotation2d(getAngleDegrees()), getLeftDistancePassedMeters(), getRightDistancePassedMeters());
+
+        SmartDashboard.putNumber("angleOfBot", getAngleDegrees());
+        SmartDashboard.putNumber("DriveLeftDistance", getLeftDistancePassedMeters());
+        SmartDashboard.putNumber("DriveRightDistance", getRightDistancePassedMeters());
+        SmartDashboard.putNumber("X:", differentialDriveOdometry.getPoseMeters().getX());
+        SmartDashboard.putNumber("Y:", differentialDriveOdometry.getPoseMeters().getY());
+        SmartDashboard.putData("field: ", field2d);
```

Comment:
don't display the field again and again with this. Only call this line once in the ctor. That's enough. Otherwise you just create garbage calls

Reply:
done

Reply:
What `putData` does is take an object (of type `Sendable') and store it. Then it periodically updates information from that object to shuffleboard (or from shuffleboard to the object). 

As such, unlike `putNumber`, it should really only be called once and not more.

Reply:
I understood and changed it in the code the first time you commented, and put the field2d in the ctor.
I dont know why its still like that